http/http_client.h

```diff
@@ -0,0 +1,432 @@
+/****************************************************************************/
+
+struct HttpClientImpl : public AsyncEventSource {
+    HttpClientImpl(const std::string & baseUrl,
+                   int numParallel = 1024, int queueSize = 0)
+        : AsyncEventSource()
+    {
+    }
+
+    HttpClientImpl(HttpClientImpl && other) = default;
+
+    virtual ~HttpClientImpl()
+    {}
+
+    /** Enable debugging */
+    virtual void enableDebug(bool value) = 0;
+
+    /** SSL checks */
+    virtual void enableSSLChecks(bool value) = 0;
+
+    /** Enable the TCP_NODELAY option, also known as the Nagle's algorithm */
+    virtual void enableTcpNoDelay(bool value) = 0;
+
+    /** Use with servers that support HTTP pipelining */
+    virtual void enablePipelining(bool value) = 0;
+
+    /** Enqueue (or perform) the specified request */
+    virtual bool enqueueRequest(const std::string & verb,
+                                const std::string & resource,
+                                const std::shared_ptr<HttpClientCallbacks> & callbacks,
+                                const HttpRequest::Content & content,
+                                const RestParams & queryParams,
+                                const RestParams & headers,
+                                int timeout = -1) = 0;
+
+    /* Returns the number of requests in the queue */
+    virtual size_t queuedRequests() const = 0;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT ERROR                                                        */
+/****************************************************************************/
+
+enum struct HttpClientError {
+    None,
+    Unknown,
+    Timeout,
+    HostNotFound,
+    CouldNotConnect,
+    SendError,
+    RecvError
+};
+
+std::ostream & operator << (std::ostream & stream, HttpClientError error);
+
+
+/****************************************************************************/
+/* HTTP CLIENT                                                              */
+/****************************************************************************/
+
+struct HttpClient : public AsyncEventSource {
+    /* This sets the requested version of the underlying HttpClientImpl. By
+     * default, this value is deduced from the "HTTP_CLIENT_VERSION"
+     * environment variable. It not set, this falls back to 1. */
+    static void setHttpClientImplVersion(int version);
+
+    /* "baseUrl": scheme, hostname and port (scheme://hostname[:port]) that
+       will be used as base for all requests
+       "numParallels": number of requests that can be handled simultaneously
+       "queueSize": size of the backlog of pending requests, after which
+       operations will be refused (0 = infinite)
+       "implVersion": use version X of the HttpClientImpl, fallback
+       to HTTP_CLIENT_IMPL
+    */
+    HttpClient(const std::string & baseUrl,
```

Comment:
Done. I have segregated most classes in their own files and put all the code in the .cc. Should make things much clearer and reduce the number of required includes.